src/shared/Core/GenericHostProvider.cs

```diff
@@ -86,10 +109,94 @@ public override async Task<ICredential> GenerateCredentialAsync(InputArguments i
                 Context.Trace.WriteLine("Windows Integrated Authentication detection has been disabled.");
             }
 
+            // Use basic authentication
             Context.Trace.WriteLine("Prompting for basic credentials...");
             return await _basicAuth.GetCredentialsAsync(uri.AbsoluteUri, input.UserName);
         }
 
+        private async Task<ICredential> GetOAuthAccessToken(Uri remoteUri, string userName, GenericOAuthConfig config)
+        {
+            // TODO: Determined user info from a webcall? ID token? Need OIDC support
+            string oauthUser = userName ?? config.DefaultUserName;
+
+            var client = new OAuth2Client(
+                HttpClient,
+                config.Endpoints,
+                config.ClientId,
+                config.RedirectUri,
+                config.ClientSecret,
+                Context.Trace,
+                config.UseAuthHeader);
+
+            string refreshService = remoteUri.AbsoluteUri.TrimEnd('/') + "/refresh_token";

```

Comment:
I know there are some concerns about how we generate a service name for the refresh tokens with collisions with other credentials.

@hickford, you had suggested in the past using a subdomain to differentiate ATs and RTs:

```text
https://foo.example.com/path/repo
https://refresh_token.foo.example.com/path/repo
```

Did you or @ldennington have any other suggestions? I was going to change to the `refresh_token` subdomain example above if there was no objections.

Reply:
I'm fine with the subdomain implementation.